File: shell/design.py
"""Design add-on — capability, tokeni i override pohrana."""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any

from shell.platform import load_shell_config

_APP_ROOT = Path(__file__).resolve().parent.parent
_TOKENS_PATH = _APP_ROOT / "plugins" / "design-tools" / "design" / "tokens.json"
_OVERRIDES_DIR = _APP_ROOT / "data" / "design_overrides"
_TOKEN_OVERRIDES_PATH = _OVERRIDES_DIR / "tokens.json"
_THEMES_DIR = _OVERRIDES_DIR / "themes"
_ACTIVE_THEME_PATH = _OVERRIDES_DIR / "active_theme.json"
_DEFAULT_THEME_ID = "default"

_VALID_MODES = frozenset({"open", "password", "off"})


def design_mode() -> str:
    env = os.environ.get("QNC_DESIGN_MODE", "").strip().lower()
    if env in _VALID_MODES:
        return env
    cfg = load_shell_config()
    block = cfg.get("design_editor")
    if isinstance(block, dict):
        mode = str(block.get("mode") or "").strip().lower()
        if mode in _VALID_MODES:
            return mode
    return "open"


def design_default_enabled() -> bool:
    cfg = load_shell_config()
    block = cfg.get("design_editor")
    if isinstance(block, dict) and "default_enabled" in block:
        return bool(block.get("default_enabled"))
    return design_mode() == "open"


def design_editor_capability() -> dict[str, Any]:
    mode = design_mode()
    available = mode != "off"
    return {
        "available": available,
        "mode": mode,
        "authenticated": available and mode == "open",
        "default_enabled": design_default_enabled(),
    }


def _read_json(path: Path) -> dict[str, Any]:
    if not path.is_file():
        return {}
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return {}
    return data if isinstance(data, dict) else {}


def _write_json(path: Path, data: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(data, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


def load_base_tokens() -> dict[str, Any]:
    return _read_json(_TOKENS_PATH)


def _slug_theme_id(label: str) -> str:
    import re

    slug = re.sub(r"[^a-z0-9]+", "-", label.strip().lower()).strip("-")
    return slug or "tema"


def active_theme_id() -> str:
    data = _read_json(_ACTIVE_THEME_PATH)
    theme_id = str(data.get("theme_id") or _DEFAULT_THEME_ID).strip()
    return theme_id or _DEFAULT_THEME_ID


def _theme_path(theme_id: str) -> Path:
    return _THEMES_DIR / f"{theme_id}.json"


def _load_theme_doc(theme_id: str) -> dict[str, Any]:
    if theme_id == _DEFAULT_THEME_ID:
        return {
            "id": _DEFAULT_THEME_ID,
            "label": load_base_tokens().get("label", "QNC Default"),
            "built_in": True,
            "tokens": load_token_overrides(),
        }
    return _read_json(_theme_path(theme_id))


def list_themes() -> dict[str, Any]:
    base = load_base_tokens()
    themes: list[dict[str, Any]] = [
        {
            "id": _DEFAULT_THEME_ID,
            "label": str(base.get("label") or "QNC Default"),
            "built_in": True,
        }
    ]
    if _THEMES_DIR.is_dir():
        for path in sorted(_THEMES_DIR.glob("*.json")):
            doc = _read_json(path)
            theme_id = str(doc.get("id") or path.stem).strip()
            if not theme_id or theme_id == _DEFAULT_THEME_ID:
                continue
            themes.append(
                {
                    "id": theme_id,
                    "label": str(doc.get("label") or theme_id),
                    "built_in": False,
                }
            )
    active = active_theme_id()
    return {"status": "ok", "active_id": active, "themes": themes}


def load_token_overrides() -> dict[str, str]:
    data = _read_json(_TOKEN_OVERRIDES_PATH)
    tokens = data.get("tokens")
    if not isinstance(tokens, dict):
        return {}
    return {str(k): str(v) for k, v in tokens.items() if str(k).startswith("--")}


def _theme_override_tokens(theme_id: str) -> dict[str, str]:
    if theme_id == _DEFAULT_THEME_ID:
        return load_token_overrides()
    doc = _load_theme_doc(theme_id)
    raw = doc.get("tokens")
    if not isinstance(raw, dict):
        return {}
    return {str(k): str(v) for k, v in raw.items() if str(k).startswith("--")}


def merged_tokens() -> dict[str, Any]:
    base = load_base_tokens()
    theme_id = active_theme_id()
    overrides = _theme_override_tokens(theme_id)
    tokens = dict(base.get("tokens") or {})
    tokens.update(overrides)
    theme_doc = _load_theme_doc(theme_id)
    label = str(theme_doc.get("label") or base.get("label") or "QNC Default")
    return {
        "status": "ok",
        "version": base.get("version", 1),
        "label": label,
        "theme_id": theme_id,
        "tokens": tokens,
        "overrides": overrides,
    }


def save_token_overrides(tokens: dict[str, str], theme_id: str | None = None) -> dict[str, str]:
    clean = {str(k): str(v) for k, v in tokens.items() if str(k).startswith("--")}
    target = (theme_id or active_theme_id()).strip() or _DEFAULT_THEME_ID
    if target == _DEFAULT_THEME_ID:
        _write_json(_TOKEN_OVERRIDES_PATH, {"version": 1, "tokens": clean})
        return clean
    path = _theme_path(target)
    if not path.is_file():
        raise ValueError(f"Tema '{target}' ne postoji.")
    doc = _read_json(path)
    _write_json(
        _theme_path(target),
        {
            "version": 1,
            "id": target,
            "label": str(doc.get("label") or target),
            "tokens": clean,
        },
    )
    return clean


def create_theme(label: str) -> dict[str, Any]:
    clean_label = label.strip()
    if not clean_label:
        raise ValueError("Naziv teme je obavezan.")
    theme_id = _slug_theme_id(clean_label)
    suffix = 2
    while _theme_path(theme_id).is_file():
        theme_id = f"{_slug_theme_id(clean_label)}-{suffix}"
        suffix += 1
    snapshot = merged_tokens().get("tokens") or {}
    base = load_base_tokens().get("tokens") or {}
    overrides = {
        str(k): str(v)
        for k, v in snapshot.items()
        if str(k).startswith("--") and str(v) != str(base.get(k, ""))
    }
    _THEMES_DIR.mkdir(parents=True, exist_ok=True)
    _write_json(
        _theme_path(theme_id),
        {"version": 1, "id": theme_id, "label": clean_label, "tokens": overrides},
    )
    activate_theme(theme_id)
    return {"status": "ok", "id": theme_id, "label": clean_label, "active_id": theme_id}


def activate_theme(theme_id: str) -> dict[str, Any]:
    target = theme_id.strip() or _DEFAULT_THEME_ID
    if target != _DEFAULT_THEME_ID and not _theme_path(target).is_file():
        raise ValueError(f"Tema '{target}' ne postoji.")
    _write_json(_ACTIVE_THEME_PATH, {"theme_id": target})
    return {"status": "ok", "active_id": target}


def design_status() -> dict[str, Any]:
    cap = design_editor_capability()
    return {
        "status": "ok",
        **cap,
        "paths": {
            "tokens": "/plugins/design-tools/design/tokens.json",
            "overrides": "/data/design_overrides/tokens.json",
        },
    }

File: qnc-host/src/design.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};

use rusqlite::Connection;
use serde_json::{json, Value};

use crate::config::{read_json, AppConfig};
use crate::design_db::{
    self, active_theme_id as db_active_theme_id, lab_key, load_custom_theme_ids, load_json_setting,
    save_custom_theme_ids, save_json_setting, set_active_theme_id, theme_doc_key, theme_exists,
    DEFAULT_TOKENS_KEY,
};
use crate::project::db::{open_global, ProjectPaths};

const VALID_MODES: [&str; 3] = ["open", "password", "off"];
const DEFAULT_THEME_ID: &str = "default";

fn with_design_store<T>(
    root: &Path,
    config: &AppConfig,
    f: impl FnOnce(&Connection, &Path) -> Result<T, String>,
) -> Result<T, String> {
    let paths = ProjectPaths::from_root(root, config);
    let conn = open_global(&paths).map_err(|e| e.to_string())?;
    design_db::ensure_files_migrated(&conn, root)?;
    f(&conn, root)
}

pub fn design_mode(root: &Path) -> String {
    if let Ok(env) = std::env::var("QNC_DESIGN_MODE") {
        let mode = env.trim().to_lowercase();
        if VALID_MODES.contains(&mode.as_str()) {
            return mode;
        }
    }
    let cfg = read_json(&root.join("data").join("shell_config.json")).unwrap_or(json!({}));
    if let Some(mode) = cfg
        .get("design_editor")
        .and_then(|v| v.get("mode"))
        .and_then(|v| v.as_str())
    {
        let mode = mode.trim().to_lowercase();
        if VALID_MODES.contains(&mode.as_str()) {
            return mode;
        }
    }
    "open".to_string()
}

pub fn design_default_enabled(root: &Path, mode: &str) -> bool {
    let cfg = read_json(&root.join("data").join("shell_config.json")).unwrap_or(json!({}));
    if let Some(block) = cfg.get("design_editor") {
        if let Some(v) = block.get("default_enabled") {
            return v.as_bool().unwrap_or(false);
        }
    }
    mode == "open"
}

pub fn design_editor_capability(root: &Path) -> Value {
    let mode = design_mode(root);
    let available = mode != "off";
    let authenticated = available && mode == "open";
    json!({
        "available": available,
        "mode": mode,
        "authenticated": authenticated,
        "default_enabled": design_default_enabled(root, &mode),
    })
}

pub fn design_status(root: &Path, config: &AppConfig) -> Value {
    let cap = design_editor_capability(root);
    let _ = with_design_store(root, config, |_, _| Ok(()));
    json!({
        "status": "ok",
        "available": cap.get("available").cloned().unwrap_or(json!(false)),
        "mode": cap.get("mode").cloned().unwrap_or(json!("off")),
        "authenticated": cap.get("authenticated").cloned().unwrap_or(json!(false)),
        "default_enabled": cap.get("default_enabled").cloned().unwrap_or(json!(false)),
        "storage": "app_settings",
        "paths": {
            "tokens": "/plugins/design-tools/design/tokens.json",
            "overrides": "project_store.db/app_settings (design.*)",
            "themes": "project_store.db/app_settings (design.theme.*)",
            "active_theme": "project_store.db/app_settings (design.active_theme_id)",
            "timeline_lab": "project_store.db/app_settings (design.lab.timeline)",
            "project_list_lab": "project_store.db/app_settings (design.lab.project_list)",
            "project_template_settings_lab": "project_store.db/app_settings (design.lab.project_template_settings)",
            "ingest_clip_grid_lab": "project_store.db/app_settings (design.lab.ingest_clip_grid)",
        }
    })
}

fn slug_theme_id(label: &str) -> String {
    let mut slug = String::new();
    let mut prev_dash = false;
    for ch in label.trim().to_lowercase().chars() {
        if ch.is_ascii_alphanumeric() {
            slug.push(ch);
            prev_dash = false;
        } else if !prev_dash && !slug.is_empty() {
            slug.push('-');
            prev_dash = true;
        }
    }
    let slug = slug.trim_matches('-').to_string();
    if slug.is_empty() {
        "tema".to_string()
    } else {
        slug
    }
}

fn base_tokens_path(root: &Path) -> PathBuf {
    root.join("plugins")
        .join("design-tools")
        .join("design")
        .join("tokens.json")
}

fn load_token_overrides(conn: &Connection) -> HashMap<String, String> {
    let doc = load_json_setting(conn, DEFAULT_TOKENS_KEY)
        .ok()
        .flatten()
        .unwrap_or(json!({}));
    token_map_from_doc(&doc)
}

fn token_map_from_doc(doc: &Value) -> HashMap<String, String> {
    let mut out = HashMap::new();
    if let Some(obj) = doc.get("tokens").and_then(|v| v.as_object()) {
        for (k, v) in obj {
            if k.starts_with("--") {
                if let Some(s) = v.as_str() {
                    out.insert(k.clone(), s.to_string());
                }
            }
        }
    }
    out
}

fn load_theme_doc(conn: &Connection, root: &Path, theme_id: &str) -> Value {
    if theme_id == DEFAULT_THEME_ID {
        let base = read_json(&base_tokens_path(root)).unwrap_or(json!({}));
        let overrides = load_token_overrides(conn);
        return json!({
            "id": DEFAULT_THEME_ID,
            "label": base.get("label").cloned().unwrap_or(json!("QNC Default")),
            "built_in": true,
            "tokens": overrides,
        });
    }
    load_json_setting(conn, &theme_doc_key(theme_id))
        .ok()
        .flatten()
        .unwrap_or(json!({}))
}

fn theme_override_tokens(
    conn: &Connection,
    root: &Path,
    theme_id: &str,
) -> HashMap<String, String> {
    if theme_id == DEFAULT_THEME_ID {
        return load_token_overrides(conn);
    }
    let doc = load_theme_doc(conn, root, theme_id);
    token_map_from_doc(&json!({ "tokens": doc.get("tokens").cloned().unwrap_or(json!({})) }))
}

pub fn merged_tokens(root: &Path, config: &AppConfig) -> Value {
    with_design_store(root, config, |conn, root| {
        let base =
            read_json(&base_tokens_path(root)).unwrap_or(json!({ "version": 1, "tokens": {} }));
        let theme_id = db_active_theme_id(conn)?;
        let overrides_map = theme_override_tokens(conn, root, &theme_id);
        let mut tokens = HashMap::new();
        if let Some(obj) = base.get("tokens").and_then(|v| v.as_object()) {
            for (k, v) in obj {
                if let Some(s) = v.as_str() {
                    tokens.insert(k.clone(), s.to_string());
                }
            }
        }
        for (k, v) in &overrides_map {
            tokens.insert(k.clone(), v.clone());
        }
        let theme_doc = load_theme_doc(conn, root, &theme_id);
        let label = theme_doc
            .get("label")
            .and_then(|v| v.as_str())
            .unwrap_or("QNC Default");
        let overrides: Value = serde_json::to_value(&overrides_map).unwrap_or(json!({}));
        Ok(json!({
            "status": "ok",
            "version": base.get("version").cloned().unwrap_or(json!(1)),
            "label": label,
            "theme_id": theme_id,
            "tokens": tokens,
            "overrides": overrides,
        }))
    })
    .unwrap_or_else(|_| json!({ "status": "error" }))
}

pub fn list_themes(root: &Path, config: &AppConfig) -> Value {
    with_design_store(root, config, |conn, root| {
        let base = read_json(&base_tokens_path(root)).unwrap_or(json!({}));
        let mut themes = vec![json!({
            "id": DEFAULT_THEME_ID,
            "label": base.get("label").cloned().unwrap_or(json!("QNC Default")),
            "built_in": true,
        })];
        for theme_id in load_custom_theme_ids(conn)? {
            if theme_id == DEFAULT_THEME_ID {
                continue;
            }
            let doc = load_theme_doc(conn, root, &theme_id);
            themes.push(json!({
                "id": theme_id,
                "label": doc.get("label").cloned().unwrap_or(json!(theme_id)),
                "built_in": false,
            }));
        }
        Ok(json!({
            "status": "ok",
            "active_id": db_active_theme_id(conn)?,
            "themes": themes,
        }))
    })
    .unwrap_or_else(|_| json!({ "status": "error", "themes": [] }))
}

pub fn activate_theme(root: &Path, config: &AppConfig, theme_id: &str) -> Result<Value, String> {
    let target = if theme_id.trim().is_empty() {
        DEFAULT_THEME_ID.to_string()
    } else {
        theme_id.trim().to_string()
    };
    with_design_store(root, config, |conn, _| {
        if target != DEFAULT_THEME_ID && !theme_exists(conn, &target)? {
            return Err(format!("Tema '{target}' ne postoji."));
        }
        set_active_theme_id(conn, &target)?;
        Ok(json!({ "status": "ok", "active_id": target }))
    })
}

pub fn create_theme(root: &Path, config: &AppConfig, label: &str) -> Result<Value, String> {
    let clean = label.trim();
    if clean.is_empty() {
        return Err("Naziv teme je obavezan.".into());
    }
    with_design_store(root, config, |conn, root| {
        let mut theme_id = slug_theme_id(clean);
        let mut suffix = 2;
        while theme_exists(conn, &theme_id)? {
            theme_id = format!("{}-{}", slug_theme_id(clean), suffix);
            suffix += 1;
        }
        let merged = merged_tokens(root, config);
        let base = read_json(&base_tokens_path(root)).unwrap_or(json!({}));
        let base_tokens = base.get("tokens").and_then(|v| v.as_object());
        let mut overrides = HashMap::new();
        if let Some(obj) = merged.get("tokens").and_then(|v| v.as_object()) {
            for (k, v) in obj {
                if !k.starts_with("--") {
                    continue;
                }
                let val = v.as_str().unwrap_or("");
                let base_val = base_tokens
                    .and_then(|b| b.get(k))
                    .and_then(|v| v.as_str())
                    .unwrap_or("");
                if val != base_val {
                    overrides.insert(k.clone(), val.to_string());
                }
            }
        }
        let doc = json!({
            "version": 1,
            "id": theme_id,
            "label": clean,
            "tokens": overrides,
        });
        save_json_setting(conn, &theme_doc_key(&theme_id), &doc)?;
        let mut ids = load_custom_theme_ids(conn)?;
        if !ids.iter().any(|id| id == &theme_id) {
            ids.push(theme_id.clone());
            save_custom_theme_ids(conn, &ids)?;
        }
        set_active_theme_id(conn, &theme_id)?;
        Ok(json!({
            "status": "ok",
            "id": theme_id.clone(),
            "label": clean,
            "active_id": theme_id,
        }))
    })
}

pub fn save_token_overrides(
    root: &Path,
    config: &AppConfig,
    tokens: &HashMap<String, String>,
) -> Result<Value, String> {
    let mode = design_mode(root);
    if mode == "off" {
        return Err("Design editor je isključen.".into());
    }
    if mode == "password" {
        return Err("Spremanje zahtijeva admin autentifikaciju (još nije implementirano).".into());
    }
    with_design_store(root, config, |conn, root| {
        let target = db_active_theme_id(conn)?;
        let clean: HashMap<String, String> = tokens
            .iter()
            .filter(|(k, _)| k.starts_with("--"))
            .map(|(k, v)| (k.clone(), v.clone()))
            .collect();
        if target == DEFAULT_THEME_ID {
            let doc = json!({ "version": 1, "tokens": clean });
            save_json_setting(conn, DEFAULT_TOKENS_KEY, &doc)?;
            return Ok(json!({ "status": "ok", "theme_id": target, "tokens": clean }));
        }
        let theme_doc = load_theme_doc(conn, root, &target);
        if theme_doc.get("id").is_none() {
            return Err(format!("Tema '{target}' ne postoji."));
        }
        let label = theme_doc
            .get("label")
            .and_then(|v| v.as_str())
            .unwrap_or(&target);
        let doc = json!({
            "version": 1,
            "id": target,
            "label": label,
            "tokens": clean,
        });
        save_json_setting(conn, &theme_doc_key(&target), &doc)?;
        Ok(json!({ "status": "ok", "theme_id": target, "tokens": clean }))
    })
}

fn default_timeline_lab_prefs() -> Value {
    json!({
        "version": 1,
        "active_profile_id": "design-lab",
        "label_display": "full",
        "video_presentation": "filmstrip",
        "playhead_pct": 35,
        "track_states": {},
    })
}

pub fn load_timeline_lab_prefs(root: &Path, config: &AppConfig) -> Value {
    with_design_store(root, config, |conn, _| {
        let doc = load_json_setting(conn, &lab_key("timeline"))?
            .unwrap_or_else(default_timeline_lab_prefs);
        Ok(json!({
            "status": "ok",
            "prefs": doc,
        }))
    })
    .unwrap_or_else(|_| json!({ "status": "error" }))
}

pub fn save_timeline_lab_prefs(
    root: &Path,
    config: &AppConfig,
    prefs: Value,
) -> Result<Value, String> {
    let mode = design_mode(root);
    if mode == "off" {
        return Err("Design editor je isključen.".into());
    }
    if mode == "password" {
        return Err("Spremanje zahtijeva admin autentifikaciju (još nije implementirano).".into());
    }
    with_design_store(root, config, |conn, _| {
        save_json_setting(conn, &lab_key("timeline"), &prefs)?;
        Ok(json!({ "status": "ok", "prefs": prefs }))
    })
}

fn default_project_list_lab_prefs() -> Value {
    json!({
        "version": 1,
        "component": "project-list",
        "active_profile_id": "design-lab",
        "feature_states": {},
    })
}

pub fn load_project_list_lab_prefs(root: &Path, config: &AppConfig) -> Value {
    with_design_store(root, config, |conn, _| {
        let doc = load_json_setting(conn, &lab_key("project_list"))?
            .unwrap_or_else(default_project_list_lab_prefs);
        Ok(json!({
            "status": "ok",
            "prefs": doc,
        }))
    })
    .unwrap_or_else(|_| json!({ "status": "error" }))
}

pub fn save_project_list_lab_prefs(
    root: &Path,
    config: &AppConfig,
    prefs: Value,
) -> Result<Value, String> {
    let mode = design_mode(root);
    if mode == "off" {
        return Err("Design editor je isključen.".into());
    }
    if mode == "password" {
        return Err("Spremanje zahtijeva admin autentifikaciju (još nije implementirano).".into());
    }
    with_design_store(root, config, |conn, _| {
        save_json_setting(conn, &lab_key("project_list"), &prefs)?;
        Ok(json!({ "status": "ok", "prefs": prefs }))
    })
}

fn default_project_template_settings_lab_prefs() -> Value {
    json!({
        "version": 1,
        "component": "project-template-settings",
        "active_profile_id": "design-lab",
        "section_states": {},
    })
}

pub fn load_project_template_settings_lab_prefs(root: &Path, config: &AppConfig) -> Value {
    with_design_store(root, config, |conn, _| {
        let doc = load_json_setting(conn, &lab_key("project_template_settings"))?
            .unwrap_or_else(default_project_template_settings_lab_prefs);
        Ok(json!({
            "status": "ok",
            "prefs": doc,
        }))
    })
    .unwrap_or_else(|_| json!({ "status": "error" }))
}

pub fn save_project_template_settings_lab_prefs(
    root: &Path,
    config: &AppConfig,
    prefs: Value,
) -> Result<Value, String> {
    let mode = design_mode(root);
    if mode == "off" {
        return Err("Design editor je isključen.".into());
    }
    if mode == "password" {
        return Err("Spremanje zahtijeva admin autentifikaciju (još nije implementirano).".into());
    }
    with_design_store(root, config, |conn, _| {
        save_json_setting(conn, &lab_key("project_template_settings"), &prefs)?;
        Ok(json!({ "status": "ok", "prefs": prefs }))
    })
}

fn default_ingest_clip_grid_lab_prefs() -> Value {
    json!({
        "version": 1,
        "component": "ingest-clip-grid",
        "active_profile_id": "fcp11-browser",
        "feature_states": {},
    })
}

pub fn load_ingest_clip_grid_lab_prefs(root: &Path, config: &AppConfig) -> Value {
    with_design_store(root, config, |conn, _| {
        let doc = load_json_setting(conn, &lab_key("ingest_clip_grid"))?
            .unwrap_or_else(default_ingest_clip_grid_lab_prefs);
        Ok(json!({
            "status": "ok",
            "prefs": doc,
        }))
    })
    .unwrap_or_else(|_| json!({ "status": "error" }))
}

pub fn save_ingest_clip_grid_lab_prefs(
    root: &Path,
    config: &AppConfig,
    prefs: Value,
) -> Result<Value, String> {
    let mode = design_mode(root);
    if mode == "off" {
        return Err("Design editor je isključen.".into());
    }
    if mode == "password" {
        return Err("Spremanje zahtijeva admin autentifikaciju (još nije implementirano).".into());
    }
    with_design_store(root, config, |conn, _| {
        save_json_setting(conn, &lab_key("ingest_clip_grid"), &prefs)?;
        Ok(json!({ "status": "ok", "prefs": prefs }))
    })
}
